pkg/components/ebpf/common/kafkaparser/common.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	if header.MessageSize > KafkaMaxPayloadLen {
+		return errors.New("invalid Kafka response header: message size exceeds maximum payload length")
+	}
+
+	if header.CorrelationID < 0 {
+		return errors.New("invalid Kafka request header: correlation ID is negative")
+	}
+	if header.CorrelationID != requestHeader.CorrelationID {
+		return errors.New("invalid Kafka response header: correlation ID does not match request header")
+	}
+	return nil
+}
+
+// isFlexible checks for each API key if the version is flexible.
+// a flexible version uses a dynamic size for arrays and strings
+func isFlexible(header *KafkaRequestHeader) bool {
+	switch header.APIKey {
+	// https://github.com/apache/kafka/blob/9983331d917fe8f57c37c88f0749b757e5af0c87/clients/src/main/resources/common/message/ProduceRequest.json#L51
+	case APIKeyProduce:
+		return header.APIVersion >= 9
+	// https://github.com/apache/kafka/blob/9983331d917fe8f57c37c88f0749b757e5af0c87/clients/src/main/resources/common/message/FetchRequest.json#L62C4-L62C20
+	case APIKeyFetch:
+		return header.APIVersion >= 12
+	// https://github.com/apache/kafka/blob/9983331d917fe8f57c37c88f0749b757e5af0c87/clients/src/main/resources/common/message/MetadataRequest.json#L22
+	case APIKeyMetadata:
+		return header.APIVersion >= 9
+	default:
+		return false
+	}
+}
+
+func readArrayLength(pkt []byte, header *KafkaRequestHeader, offset Offset) (int, Offset, error) {
+	if isFlexible(header) {
+		size, offset, err := readUnsignedVarint(pkt[offset:], offset)
+		if size == 0 {
+			return 0, offset, nil // return 0 for null
+		}
+		return size - 1, offset, err
+	} else {
+		return readInt32(pkt, offset)
+	}
+}
+
+func readUUID(pkt []byte, offset Offset) (*UUID, Offset, error) {
+	if offset+UUIDLen > len(pkt) {
+		return nil, offset, errors.New("packet too short for topic UUID")
+	}
+	uuid := (UUID)(pkt[offset : offset+UUIDLen])
+	return &uuid, offset + UUIDLen, nil
+}
+
+func readString(pkt []byte, header *KafkaRequestHeader, offset Offset, nullable bool) (string, Offset, error) {
+	size, offset, err := readStringLength(pkt, header, offset, nullable)
+	if err != nil {
+		return "", offset, err
+	}
+	if nullable && size == 0 {
+		return "", offset, nil // return empty string for null
+	}
+	if offset+size > len(pkt) {
+		return "", 0, errors.New("string size exceeds packet size")
+	}
+	str := string(pkt[offset : offset+size])
+	return str, offset + size, nil
+}
+
+func readStringLength(pkt []byte, header *KafkaRequestHeader, offset Offset, nullable bool) (int, Offset, error) {
+	if !isFlexible(header) {
+		// length is stored as a fixed size int16
+		if offset+Int16Len > len(pkt) {
+			return 0, 0, errors.New("packet too short for string length")
+		}
+		size := int16(binary.BigEndian.Uint16(pkt[offset:]))
+		if nullable && size == -1 {
+			return 0, offset + Int16Len, nil // return 0 for null
+		}
+		return int(size), offset + Int16Len, nil // size is stored as a varint, so we subtract 1
```

Comment:
is the comment outdated? I don't see the subtraction

Reply:
yeah is a couple rows down :), will remove